File: 2024/03Mar/gfg/29_euler-circuit-in-a-directed-graph.java
import java.util.*;

class Solution {
    public boolean isEularCircuitExist(int v, ArrayList<ArrayList<Integer>> adj) {
        // Code here
        HashSet<Integer> visited = new HashSet<Integer>();
        HashSet<Integer> cur = new HashSet<Integer>();
        
        return dfs(0, visited, cur, -1, adj);
    }
    
    public static boolean dfs(int node, HashSet<Integer> visited, HashSet<Integer> cur, int parent, ArrayList<ArrayList<Integer>> adj) {
        if(cur.contains(node)) return true;
        if(visited.contains(node)) return false;
        
        visited.add(node);
        
        ArrayList<Integer> list = adj.get(node);
        
        cur.add(node);
        for(int i=0;i<list.size();i++) {
            if(list.get(i) == parent) continue;
            if(dfs(list.get(i), visited, cur, node, adj)) return true;
        }
        cur.remove(node);
        
        return false;
    }
}

// not working for some cases

File: 2024/03Mar/gfg/29_euler-circuit-in-a-directed-graph.js
class Solution {
    isEularCircuitExist(v, adj){
       //code here
       const visited = new Array(v).fill(false);
       const cur = new Array(v).fill(false);
       
       return dfs(0, -1);
       
       function dfs(node, parent) {
           if(cur[node]) return true;
           if(visited[node]) return false;
           
           visited[node] = true;
           cur[node] = true;
           
           for(let e of adj[node]) {
               if(e == parent) continue;
               if(dfs(e, node)) return true;
           }
           
           cur[node] = false;
           return false;
       }
    }
}

// Not working for some cases